llama: extend for small granite models

giuseppe opened the pull request:
it works only for the small models 3b and 8b.  The bigger models work fine with the existing GPTBigCodeForCausalLM architecture.

For the small models there are enough differences with the base llama arch that it is worth to define a new architecture.

To create the .gguf files, it is necessary to specify GraniteSmallForCausalLM in the architectures for the hf model.

Closes: https://github.com/ggerganov/llama.cpp/issues/7116

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

sroecker committed:
Add optional MLP bias for Granite models

Add optional MLP bias for ARCH_LLAMA to support Granite models.
Partially addresses ggerganov/llama.cpp/issues/7116
Still needs some more changes to properly support Granite.

giuseppe committed:
llama: add support for small granite models

it works only for the small models 3b and 8b.

The convert-hf-to-gguf.py script uses the vocabulary size of the
granite models to detect granite and set the correct configuration.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

ggerganov commented on llama.cpp:
```diff
@@ -7133,9 +7183,9 @@ struct llm_build_context {
-                        model.layers[il].ffn_up,   NULL,
-                        model.layers[il].ffn_gate, NULL,
-                        model.layers[il].ffn_down, NULL,
+                        model.layers[il].ffn_up,   model.layers[il].ffn_up_b,
+                        model.layers[il].ffn_gate, model.layers[il].ffn_gate_b,
+                        model.layers[il].ffn_down, model.layers[il].ffn_down_b,
```
Initialize the bias tensors with `nullptr` in the struct declaration

giuseppe replied:
fixed, thanks

ggerganov commented on llama.cpp:
```diff
@@ -4396,6 +4437,9 @@ static void llm_load_vocab(
         } else {
             if (tokenizer_model == "gpt2") {
                 vocab.type = LLAMA_VOCAB_TYPE_BPE;
+                if (model.arch == LLM_ARCH_LLAMA) {
+                    vocab.add_space_prefix = false;
+                }
```
Is this needed - looks wrong?

giuseppe replied:
sorry, it should be `LLM_ARCH_GRANITE_SMALL`

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2465,6 +2465,34 @@ def set_vocab(self, *args, **kwargs):
+    def set_gguf_parameters(self):
+        super().set_gguf_parameters()
+        hparams = self.hparams
+        self.gguf_writer.add_name("GraniteSmall")
```
Leaving this here makes the model have 2 names, because `super().set_gguf_parameters()` already sets it according to the model's directory name.

```suggestion
```

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2465,6 +2465,34 @@ def set_vocab(self, *args, **kwargs):
+        self.gguf_writer.add_rope_dimension_count(hparams["hidden_size"] // hparams["num_attention_heads"])
+        self.gguf_writer.add_add_bos_token(False)
+        self.gguf_writer.add_rope_scaling_type(gguf.RopeScalingType.NONE)
+
+    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
+        return [(self.map_tensor_name(name), data_torch)]
```
If it's the same as for the base `Model` class, the `modify_tensors` function doesn't need to be overridden.

```suggestion
```

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2465,6 +2465,34 @@ def set_vocab(self, *args, **kwargs):
+class GraniteModel(Model):
+    model_arch = gguf.MODEL_ARCH.GRANITE_SMALL
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
```
This is also not needed.

```suggestion
```

giuseppe commented:
@compilade thanks, addressed the issues and pushed a new version

ggerganov commented:
Adding the `--architecture` argument should be avoided. Instead of adding new architecture `GraniteSmallForCausalLM`, try to update the existing `class LlamaModel` to handle this model. Since it is a BPE tokenizer, you might have to update the `convert-hf-to-gguf-update.py` script as described in #6920

giuseppe commented:
> Adding the `--architecture` argument should be avoided. Instead of adding new architecture `GraniteSmallForCausalLM`, try to update the existing `class LlamaModel` to handle this model. Since it is a BPE tokenizer, you might have to update the `convert-hf-to-gguf-update.py` script as described in #6920

I've simplified the implementation and it is using the existing Llama model.  I've added a way to override the default rope type.  Now the only Granite specific code in llama.cpp is to detect `model.type`

giuseppe committed:
llama: honor add_space_prefix from the model configuration

propagate the add_space_prefix configuration from the HF model
configuration to the gguf file and honor it with the gpt2 tokenizer.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

ggerganov commented on convert-hf-to-gguf.py:
```diff
@@ -1329,10 +1342,12 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
-            data_torch = LlamaModel.permute(data_torch, n_head, n_head)
-        if name.endswith("k_proj.weight"):
-            data_torch = LlamaModel.permute(data_torch, n_head, n_kv_head)
+        # Skip for granite models
+        if self.hparams.get("vocab_size", 32000) != 49152:
+            if name.endswith("q_proj.weight"):
+                data_torch = LlamaModel.permute(data_torch, n_head, n_head)
+            if name.endswith("k_proj.weight"):
+                data_torch = LlamaModel.permute(data_torch, n_head, n_kv_head)
```
I think we can avoid adding the rope type parameter all together, by permuting the Q, K attention tensors in the correct way here. I don't have an example code unfortunately, so we need to figure out how to do it. The only difference between RoPE NORM and NEOX is that in the former we rotate the pairs `(x[2*i + 0], x[2*i + 1]`, while in the latter we rotate `(x[i], x[i + n_rot/2])`. So it's a matter of reordering the rows in each head in the correct way to make the RoPE type to be NORM - as all other LLaMA-based models

ggerganov replied:
Even though the output of `Q = rope(q)` and `K = rope(k)` would not be in the same order, it should still work because we compute `KQ = K @ Q` which is invariant to how the data in the heads is reordered - as long as it is reordered in the same way in both `K` and `Q`

I could be missing something though - not 100% confident in this. If you think it won't work, we can probably do the rope type thing, but I really prefer to find a way to avoid it

giuseppe replied:
is it something that could be changed later?

I am not confident either that it is not possible, I've spent a few hours on it and I've not been successful so far

compilade replied:
I gave this a try. Only the first `n_dims` elements of each rows should be re-ordered.

https://github.com/ggerganov/llama.cpp/blob/95f84d5ce8b449a9b16009434aca800df504a02e/ggml.c#L14418

```python3
    @staticmethod
    def permute_neox_rope(weights: Tensor, rot_dim: int) -> Tensor:
        orig_shape = weights.shape
        assert orig_shape[-1] % rot_dim == 0
        # reorder the first rot_dim elements of each row
        weights = weights.reshape((-1 , weights.shape[-1] // rot_dim, rot_dim))
        weights[:, 0, :] = weights[:, 0, :].reshape((-1, 2, rot_dim // 2)).mT.contiguous().reshape((-1, rot_dim))
        return weights.reshape((orig_shape))
```

It seems to partially work, but the output is still wrong, because in RoPE NEOX, it's only the first `rot_dim` elements per row that are roped, while in RoPE NORM, all of them are.

So it's not simply a re-ordering of elements that is necessary, unfortunately. The rope type is needed, it seems.

ggerganov replied:
Ah, true. So technically there is a difference between the 2 modes. But in this case `n_rot == n_embd_head == 80` so it should make a difference (I'm testing with `granite-3b-code-base`)

> # reorder the first rot_dim elements of each row

I'm not used to the Python syntax, so not sure but note that we have to reorder the *rows* in the weight tensors, not the elements in the rows. Later when we multiply these tensors with the hidden state, the rows "become" the elements in the row that we rope. Does that make sense?

giuseppe replied:
ah, if you are fine that I apply it directly on top of my patch then I can do.

I was thinking about you owning the ownership for the commit, since you came up with the code

giuseppe replied:
PR updated

ggerganov replied:
Nice work, thanks for looking into this!
